### ISSUE TITLE
A way to version tools without commiting all composer.* files?

### ISSUE THREAD

**Issue by Isinlor** [bamarni/composer-bin-plugin#40]
First of all, great tool, thanks for making it!

So, I would like to suggest adding a way of versioning tools without the need to commit all `composer.json` and `composer.lock` files from `./vendor-bin` in order to insure reproducible CI. Also, committing them all feels like a lot of clutter.

It would be also useful to be able to specify all tools in the main `composer.json` file. 

Currently I'm struggling with this `Makefile`:
```makefile
# runs copy paste detection
phpcpd: bin/phpcpd
	bin/phpcpd ./src --min-lines=1 --min-tokens=35

# install all bin tools
install-bin: bin/phpcpd bin/phpcbf bin/phpmd bin/phpstan bin/psalm

# install sebastian/phpcpd
bin/phpcpd: install
	composer bin phpcpd require --dev sebastian/phpcpd @stable
```

However I can't stop it from installing over and over each time I run `make phpcpd`.

**Comment by theofidry** [comment]
>So, I would like to suggest adding a way of versioning tools without the need to commit all 

And how would this be achieved? The `composer.{json|lock}` are _precisely_ to version it. If you feel it's too much clutter and don't want to look at them, what you can do however is version them as binary files, `composer.lock binary` in `.gitattributes` or something similar (you'll need to look it up).

As for your makefile issue, I would recommend the following:

```
# runs copy paste detection
phpcpd: bin/phpcpd
	bin/phpcpd ./src --min-lines=1 --min-tokens=35

# install sebastian/phpcpd
bin/phpcpd: vendor/bamarni/composer-bin-plugin
	composer bin phpcpd install sebastian/phpcpd
        touch $@

vendor/bamarni/composer-bin-plugin-plugin:
	composer install
	touch $@
```

**Comment by Isinlor** [comment]
The trick with `.gitattributes` is very cool!

I think it actually settles it for me 👍 . Thank you!

What do you think about mentioning `.gitignore` and `.gitattributes` in README tricks?

I can make pull-request.

**Comment by theofidry** [comment]
Yes it’s a good idea!

**Comment by theofidry** [comment]
Closed by #41
